integration_tests/src/main/python/cache_test.py

```diff
@@ -279,3 +279,40 @@ def helper(spark):
 
     # NOTE: we aren't comparing cpu and gpu results, we are comparing the cached and non-cached results.
     assert_equal(reg_result, cached_result)
+
+def test_function_on_cached_df(with_x_session, func, data_gen, enable_vectorized_conf):
+    def with_cache(cached):
+        def helper(spark):
+            df = unary_op_df(spark, data_gen)
+            if cached:
+                df.cache().count()
+            return func(df)
+        return helper
+
+    conf = enable_vectorized_conf.copy()
+    conf.update(allow_negative_scale_of_decimal_conf)
+    conf.update({"spark.rapids.sql.test.batchsize": "100"})
+    reg_result = with_x_session(with_cache(False), conf)
+    cached_result = with_x_session(with_cache(True), conf)
+
+    assert_equal(reg_result, cached_result)
+
+@pytest.mark.parametrize('data_gen', all_gen, ids=idfn)
+@pytest.mark.parametrize('with_x_session', [with_gpu_session, with_cpu_session])
+@pytest.mark.parametrize('enable_vectorized_conf', enable_vectorized_confs, ids=idfn)
+@ignore_order
+def test_cache_count(data_gen, with_x_session, enable_vectorized_conf):
+    test_function_on_cached_df(with_x_session, lambda df: df.count(), data_gen, enable_vectorized_conf)
+
+@pytest.mark.parametrize('data_gen', all_gen, ids=idfn)
+@pytest.mark.parametrize('with_x_session', [with_cpu_session, with_gpu_session])
+@pytest.mark.parametrize('enable_vectorized_conf', enable_vectorized_confs, ids=idfn)
+@ignore_order
+# This tests the cached and uncached values returned by collect on the CPU and GPU.
+# When running on the GPU with the DefaultCachedBatchSerializer, to project the results Spark adds a ColumnarToRowExec
+# to be able to show the results which will cause this test to throw an exception as it's not on the GPU so we have to
+# add that case to the `allowed` list. As of now there is no way for us to limit the scope of allow_non_gpu based on a
```

Comment:
So this is how I understand it, when `ParquetCachedBatchSerializer` isn't being used, the `InMemoryTableScanExec` is not on the GPU. Now if we set the `spark.sql.inMemoryColumnarStorage.enableVectorizedReader` to true, that causes the plan to use the columnar option of the serializer, that means that the output from the `InMemoryTableScanExec` is going to be columnar, therefore it needs to be converted to rows before we can collect it.

Reply:
I'm still confused here because just about every test uses `collect()` to grab the results, but I don't see most tests needing to exclude `ColumnarToRowExec` in order to perform the collect without a failure.  Do we really understand this, or was it just a workaround to get the test to pass?  Why don't other tests need this when they collect?

Reply:
Our tests don't cache, this is strictly for `InMemoryTableScanExec`. This is how I understand this. I would love to know what @revans2 thinks about my explanation